---
title: "16S Visualization"
output:
  html_document:
    toc: yes
    df_print: paged
  html_notebook:
    theme: spacelab
    toc: yes
    toc_float: yes
---

# To clear environment
rm(list = ls())

# Set up packages
dplyr -- for data manipulation -- is installed
The pipe operator %>% comes from the dplyr packages. (part of tidyverse)
dplyr cheat sheet: https://github.com/rstudio/cheatsheets/blob/main/data-transformation.pdf

```{r load_packages, message=FALSE, warning=FALSE}
# required
library(qiime2R)
library(phyloseq)
library(tidyverse)
library(vegan)

# optional
library(MetBrewer) # fun color palettes
```

# Read in data
I read in the QIIME artifacts as a phyloseq object. Phyloseq is an R package that has many convenient functions built in for manipulating and analyzing microbiome data. We will use the qiime2R package to import the QIIME data directly into a phyloseq object. 

```{r load_ps}

getwd() 

ps <- qiime2R::qza_to_phyloseq(
  features="./data/table_dada2.qza",
  tree="./data/rooted_tree.qza",
  taxonomy="./data/taxonomy.qza",
  metadata = "./data/sample-metadata.tsv" 
)

# factor the size metadata
ps@sam_data$size.mm <- factor(ps@sam_data$size.mm, levels = c("0-0.85", "0.85-1.4", "1.4-2", "2-2.8", "2.8-4", ">4"))

```

# Basic filtering and parsing
I will do a bit of manipulation using phyloseq functions, then we will use the suite of tidyverse packages to do more data manipulation. Let's start with getting relative abundance data.

```{r apply_filt}
# remove Mitochondria and Chloroplasts (removes Eukaryotes)
ps_filt0 <- phyloseq::subset_taxa(ps, ! Family %in% c("Mitochondria", "Chloroplast"))

# remove unclassified sequences
ps_filt0 <- phyloseq::subset_taxa(ps, Kingdom != "Unassigned")

# define minimum depth to rarefy
rarefy_level <- min(sample_sums(ps_filt0))
# apply rarefaction
ps_filt <-rarefy_even_depth(ps_filt0, rarefy_level,
  rngseed = 7, replace = TRUE, trimOTUs = TRUE, verbose = TRUE)

# relative abundance
ps_rel <- phyloseq::transform_sample_counts(ps_filt, function(x) x*100/sum(x))
```
# Format as data frames

This chunk saves the taxonomy, metadata, and count tables as data frames. In these tables, each row is a unique ASV. 

The pipe operator %>% comes from the dplyr packages. 

```{r reformat}
taxonomy <- as.data.frame(ps_rel@tax_table) %>% 
  rownames_to_column(var = "seq") # change the ASV ID to a column, not a row name

head(taxonomy)

metadata <- as.data.frame(as.matrix(ps_rel@sam_data)) %>% 
  rownames_to_column(var = "sample")
# factor the size metadata
metadata$size.mm <- factor(metadata$size.mm, levels = c("0-0.85", "0.85-1.4", "1.4-2", "2-2.8", "2.8-4", ">4"))

head(metadata)

table_rel <- as.data.frame(ps_rel@otu_table) %>% 
  rownames_to_column(var = "seq")

head(table_rel)


```

# Data Manipulation

This chunk manipulates the relative abundance table and combines it with the taxonomy table. We used the pivot_longer function to make this into a "long" dataframe, which is the desired format for using ggplot2. Read a bit more about long format here: http://eriqande.github.io/rep-res-web/lectures/ggplot_2_reshape_facets_stats.html. The long format will seem very chaotic, but I promise it will be extremely useful for plotting and data analysis tasks going forward.

```{r combine_mats}

table_rel_long <- table_rel %>% 
  pivot_longer(cols = !seq, names_to = "sample", values_to = "rel_ab") %>% # make a "long" dataframe
  left_join(taxonomy, join_by(seq)) %>% # join taxonomy by the sequence ID
  left_join(metadata, join_by(sample)) # join metadata by the sample ID

head(table_rel_long)
```

# Subsetting and plotting
We can use a variety of filtering functions from dplyr to manipulate data for more concise plots. For example, we could select specific bacteria to look for, or group all members of the same family or genus together. Let's do a simple example - we will plot the abundance of the genus Zoogleoa from each sampling location. First, let's see what the data looks like when we filter it to our specifications. 

Here, we define the genus we want to filter the data by.
We also order it how we want it to be displayed on the x-axis.

Adding duplicate.id to "group_by" seperates the data 
```{r sort_genus}
# make a data frame of first two lines
genus_summary <- table_rel_long %>% # perform some functions on table_rel
  group_by(Genus, size.mm, duplicate.id) %>% # group at the genus level, we will use this to sum up everything that has the same genus (ignore species)
  summarise(sum_ab = sum(rel_ab)) #%>% # sum up the relative abundance of each unique Genus at each unique combination of date and location

# Ca_Accumulibacter
gs.ca_ac <- genus_summary %>% # how to pipe data (dplyr filter)
  filter(Genus == "Ca_Accumulibacter")
show(gs.ca_ac)

# Ca_Phosphoribacter, Ca_Accumulibacter,Ca_Competibacter
gs.pac <- genus_summary %>%
  filter(Genus %in% c("Ca_Phosphoribacter", "Ca_Accumulibacter","Ca_Competibacter")) 
show(gs.pac)
```
# Scatter plots look better with duplicates

I considered making a boxplot, but I only have two data points. It is recommended to make box plots with more than 5 data points.

# Single scatter plot
```{r plot_scatter_single}
gs.ca_ac %>%
  ggplot(aes(x = size.mm, y = sum_ab)) + #, fill = duplicate.id)) +
  geom_point(
    shape = 21,
    color = "black",
    size = 3,
    stroke = 1
  ) +
  #scale_fill_manual(
    #values = c("A" = "#4E2A84", "B" = NA, "C" = "gray"),   # A = filled, B = empty
    #na.value = "white"                       # fallback fill color for NA
  #) +
  labs(title = "Ca_Accumulibacter") + #, fill = "Replicate ID") +
  xlab("Sieve Range [mm]") +
  ylab("Sum of Relative Abundance")

ggsave("./results/genus_ca_ac.png")
```

# Multiple subplots
```{r plot_scatter_multiple}
gs.pac %>%
  ggplot(aes(x = size.mm, y = sum_ab, fill = duplicate.id)) + 
  facet_wrap(~Genus, nrow = 1) +
  geom_point(
    shape = 21,
    color = "black",    # black outline
    size = 3,
    stroke = 1
  ) +
  scale_fill_manual(
    values = c("A" = "#4E2A84", "B" = NA, "C"="gray"),
    na.value = "white"
  ) +
  theme(axis.text.x = element_text(angle = 45, vjust = 0.9, hjust = 1)) +
  labs(fill = "Replicate ID") +
  xlab("Sieve Range [mm]") +
  ylab("Sum of Relative Abundance")

ggsave("./results/genus_pac.png")
```

We can save that table as a new R data frame, or we can pipe that data directly into a ggplot2 command. Notice that the timepoint column is the "character" class rather than a numeric class like a "double" Let's add a modification so the timepoint column is correctly classified.


# Alpha Diversity

```{r diversity_plot}

# Base plot (no fill here)
p <- plot_richness(ps_filt, 
                   x = "size.mm", 
                   measures = c("Chao1")) #, "Shannon"))

# Add custom geom_point with manual fill scale
p + 
  geom_point(
    aes(fill = duplicate.id),
    shape = 21,
    color = "black",
    size = 3,
    stroke = 1
  ) +
  scale_fill_manual(
    values = c("A" = "#4E2A84", "B" = NA, "C" = "gray"),
    na.value = "white"
  ) +
  labs(fill = "Replicate ID") +
  xlab("Sieve Range [mm]") +
  theme(axis.text.x = element_text(angle = 45, vjust = 0.9, hjust = 1))

ggsave("./results/diversity.png")
```

# Ordination 
tutorial: https://joey711.github.io/phyloseq/plot_ordination-examples.html

Try it here ordination B diversity
```{r ordination}
cols <- c('red','orange','green', 'blue', 'violet', 'black')

ps.ord <- ordinate(ps_filt, "NMDS", "bray")
p1 = plot_ordination(ps_filt, ps.ord, type="samples", color="size.mm", title="NMDS") + 
                     scale_color_manual(values=cols) 
  
print(p1)

ggsave("./results/ordination-NMDS.png", plot = p1, scale = 1,dpi = 300)

# ntaxa(ps_filt) # how many ASVs are there?
```
Try pcoa instead
```{r}
ps.ord <- ordinate(ps_filt, "PCoA", "unifrac")
p1 = plot_ordination(ps_filt, ps.ord, type="samples", color="size.mm", title="PCoA") + 
                     scale_color_manual(values=cols)

print(p1)
ggsave("./results/ordination-PCoA.png", plot = p1, scale = 1,dpi = 300)
```


# Distance
What is the difference between weighted and unweighted unifrac

```{r}
ps_filt.dist <- distance(ps_filt, method = "wunifrac" ) # weighted
show(ps_filt.dist)
```
R statistic ranges from -1 to 1. 0 is random. What is -1?
```{r}
ps_filt.ano <- anosim(ps_filt.dist, metadata$size.mm)
summary(ps_filt.ano)
plot(ps_filt.ano)
```

